src/node_messaging.cc

```diff
@@ -1136,14 +1136,43 @@ void JSTransferable::New(const FunctionCallbackInfo<Value>& args) {
   new JSTransferable(Environment::GetCurrent(args), args.This());
 }
 
+Local<Function> JSTransferable::GetConstructorFunction(Isolate* isolate) {
+  Local<FunctionTemplate> js_transferable_template = FunctionTemplate::New(
+      isolate,
+      New,
+      Local<Value>(),
+      Local<v8::Signature>(),
+      0,
+      v8::ConstructorBehavior::kAllow,
+      v8::SideEffectType::kHasSideEffect,
+      nullptr);
+  Local<FunctionTemplate> tmpl = FunctionTemplate::New(
+      isolate,
+      nullptr,
+      Local<Value>(),
+      Local<v8::Signature>(),
+      0,
+      v8::ConstructorBehavior::kAllow,
+      v8::SideEffectType::kHasSideEffect,
+      nullptr);
+  tmpl->SetClassName(FIXED_ONE_BYTE_STRING(isolate, "BaseObject"));
```

Comment:
```suggestion
  tmpl->SetClassName(FIXED_ONE_BYTE_STRING(isolate, "JSTransferable"));
```

??

Reply:
@jasnell this is creating the `BaseObject` constructor template, so that `JSTransferable` can inherit from it. It has been taken from here: https://github.com/nodejs/node/blob/ced5fc39303438f09f09c3036afc69456ee552bd/src/env.cc#L1716

Reply:
This is creating a new separate FunctionTemplate for BaseObject that is not the same as `BaseObject`. That seems like a breaking difference here. @addaleax ... what do you think here?

Reply:
If this is correct, comments here explaining what is happening and why would be good

Reply:
🤦 missed this. I too think this is the issue here, see - https://github.com/nodejs/node/pull/41148#discussion_r771517231. I noticed this yesterday too when I tried to locally move the same DOMException code to a context-aware module and it stopped throwing a DataCloneError. Given that the JSTransferable constructor is the only new thing coming from C++ land, I think this indeed is the root cause.

Reply:
I tried to make the BaseObject constructor available before the Environment is created in https://github.com/nodejs/node/pull/41148/commits/2e7f18e3d4edb62b35292db5f431ea1a26f2dbf0, so that JSTransferable can inherit from the same object.
Good news - postMessage() doesn't throw a DataCloneError and calls kClone
Bad news - the kDeserialize method doesn't get called and the 'message' event doesn't get fired on mc.port1 which keeps the event loop alive
However, still when I move the implementation to a context aware module, the DOMException object is sent over to port1.

Reply:
@addaleax @joyeecheung any clue what's going wrong here?